### PR TITLE
ChatGPT Agent V2

### PR DESCRIPTION
**Pull request by Future-Outlier**
## Tracking issue
https://github.com/flyteorg/flyte/issues/3936

## Why are the changes needed?
We want to support chatgpt agent and provide example for users to develop sync agent.

## What changes were proposed in this pull request?
- add openai-chatgpt plugin

### Setup process
0. Set up my local single binary environment. (use `pyflyte serve agent`)
1. Build a docker image with sync agent proto and chatgpt agent's gitsha.
2. Test the chatgpt task in both local and remote scenarios.

```yaml
tasks:
  task-plugins:
    enabled-plugins:
      - container
      - sidecar
      - K8S-ARRAY
      - agent-service
      
    default-for-task-types:
      chatgpt: agent-service
      custom_task: agent-service
      container: container
      container_array: K8S-ARRAY

plugins:
  agent-service:
    supportedTaskTypes:
      - sensor
      - spark
      - default_task
      - custom_task
      - chatgpt
      - sensor
      - airflow
    defaultAgent:
      endpoint: "dns:///localhost:8000"
      insecure: true
      timeouts:
        GetTask: 100s
      defaultTimeout: 100s
```

```Dockerfile
FROM python:3.9-slim-buster
USER root
WORKDIR /root
ENV PYTHONPATH /root
RUN apt-get update && apt-get install build-essential -y
RUN apt-get install git -y

RUN pip install -U git+https://github.com/Future-Outlier/flytekit.git@b058a985a43ffe8c27fd167d36cbf834d4c1732b#subdirectory=plugins/flytekit-openai

RUN pip install -U git+https://github.com/flyteorg/flyte.git@79534d163929734a6b8113c93bb0291ee84b9f4b#subdirectory=flyteidl

RUN pip install -U git+https://github.com/Future-Outlier/flytekit.git@b058a985a43ffe8c27fd167d36cbf834d4c1732b
```
```python
from flytekit import task, workflow
from flytekitplugins.chatgpt import ChatGPTTask, ChatGPTConfig

chatgpt_small_job = ChatGPTTask(
    name="chatgpt gpt-3.5-turbo",
    openai_organization="org-NayNG68kGnVXMJ8Ak4PMgQv7",
    chatgpt_config={
            "model": "gpt-3.5-turbo",
            "temperature": 0.7,
    },
)

chatgpt_big_job = ChatGPTTask(
    name="chatgpt gpt-4",
    openai_organization="org-NayNG68kGnVXMJ8Ak4PMgQv7",
    chatgpt_config={
            "model": "gpt-4",
            "temperature": 0.7,
    },
)

@workflow
def wf(message: str) -> str:
    message = chatgpt_small_job(message=message)
    message = chatgpt_big_job(message=message)
    return message

if __name__ == "__main__":
    print(wf(message="hi"))
```
#### Local Scenario
```bash
python chatgpt_chain.py
```
image
#### Remote Scenario
```bash
pyflyte run --remote --image futureoutlier/flytekit:chatgpt-0104 chatgpt_chain.py wf --message hi
```
image
image
image
image

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
https://github.com/flyteorg/flytekit/pull/1822
https://github.com/flyteorg/flytekit/pull/2085

### CODE REVIEW

**Comment by Future-Outlier** [comment]
image
TODO:
I need to get the input message from literal to str message.

File: flytekit/extend/backend/base_agent.py [hunk 1]
```diff
@@ -158,15 +159,27 @@ def get_agent_secret(secret_key: str) -> str:
...
+    Synchronous tasks run quickly and can return their results instantly. Sending a prompt to ChatGPT and getting a response, or retrieving some metadata from a backend system.
     """
 
-    _is_canceled = None
-    _agent = None
-    _entity = None
+    _clean_up_task: coroutine = None
+    _agent: AgentBase = None
+    _entity: PythonTask = None
+    _ctx: FlyteContext = FlyteContext.current_context()
```
Comment (pingsutw) [inline flytekit/extend/backend/base_agent.py:181]: could we not use current context here?

**Comment by Future-Outlier** [comment]
It works well.


https://github.com/flyteorg/flytekit/assets/76461262/69ce5cf3-a062-404c-97f7-457539754974

**Comment by Future-Outlier** [comment]
The process to merge master branch is too painful, I am going to open a new branch, sorry

**Pull request closed without merging**